File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/MedianOfNumberStream.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class MedianOfNumberStream {
  static class MedianOfStream {
    PriorityQueue<Integer> minHeap;
    PriorityQueue<Integer> maxHeap;

    MedianOfStream() {
      this.minHeap = new PriorityQueue<>((x, y) -> Integer.compare(x, y));
      this.maxHeap = new PriorityQueue<>((x, y) -> Integer.compare(y, x));
    }

    // O(logn) / O(n)
    public void insertNum(int num) {
      if (maxHeap.isEmpty() || maxHeap.peek() >= num) {
        maxHeap.add(num);
      } else {
        minHeap.add(num);
      }

      rebalance();
    }

    private void rebalance() {
      if (maxHeap.size() > minHeap.size() + 1) {
        minHeap.add(maxHeap.poll());
      } else if (minHeap.size() > maxHeap.size()) {
        maxHeap.add(minHeap.poll());
      }
    }

    // O(1) / O(1)
    public double findMedian() {
      if (maxHeap.size() == minHeap.size()) {
        return (double)(maxHeap.peek() + minHeap.peek()) / 2.0;
      }
      return maxHeap.peek();
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int nq = in.nextInt();
        in.nextLine();

        MedianOfStream mos = new MedianOfStream();
        while (nq-- > 0) {
          String[] cmd = in.nextLine().trim().split(" ");

          switch (cmd[0]) {
          case "insert":
            mos.insertNum(Integer.parseInt(cmd[1]));
            break;

          case "findmedian":
            System.out.printf("%.1f\n", mos.findMedian());
            break;
          }
        }
      }
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_3/median_of_number_stream.cpp
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

class MedianOfStream {
private:
  priority_queue<int> max_heap;
  priority_queue<int, vector<int>, greater<int>> min_heap;

  void rebalance() {
    if (max_heap.size() > min_heap.size() + 1) {
      min_heap.push(max_heap.top());
      max_heap.pop();
    } else if (min_heap.size() > max_heap.size()) {
      max_heap.push(min_heap.top());
      min_heap.pop();
    }
  }

public:
  // O(logn) / O(n)
  void insert_num(int num) {
    if (max_heap.empty() || num < max_heap.top()) {
      max_heap.push(num);
    } else {
      min_heap.push(num);
    }

    rebalance();
  }

  // O(1)
  double find_median() {
    if (max_heap.size() == min_heap.size()) {
      return (double)(max_heap.top() + min_heap.top()) / 2.0;
    }
    return max_heap.top();
  }
};

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  string s;

  cin >> tt;
  while (tt--) {
    cin >> n;
    cin.ignore(1, '\n');

    MedianOfStream mos;

    while (n--) {
      getline(cin, s);

      stringstream ss(s);

      string cmd;
      getline(ss, cmd, ' ');

      if (cmd == "findmedian") {
        cout << fixed << setprecision(1) << mos.find_median() << "\n";
      } else {
        string val;
        getline(ss, val, ' ');

        mos.insert_num(stoi(val));
      }
    }
  }

  return 0;
}